packages/core/src/api/apis/implementations/OAuthRequestManager/OAuthRequestManager.ts

```diff
@@ -0,0 +1,165 @@
+  }
+
+  createAuthRequester<T>(options: AuthRequesterOptions<T>): AuthRequester<T> {
+    const handler = new OAuthPendingRequests<T>();
+
+    const index = this.handlerCount;
+    this.handlerCount++;
+
+    handler.pending().subscribe({
+      next: (scopeRequest) => {
+        const newRequests = this.currentRequests.slice();
+        const request = this.makeAuthRequest(scopeRequest, options);
+        if (!request) {
+          delete newRequests[index];
+        } else {
+          newRequests[index] = request;
+        }
+        this.currentRequests = newRequests;
+        this.observers.forEach((observer) => observer.next(newRequests));
+      },
+    });
+
+    return (scopes) => {
+      return handler.request(scopes);
+    };
+  }
+
+  // Converts the pending request and popup options into a popup request that we can forward to subscribers.
+  private makeAuthRequest(
+    request: PendingRequest<any>,
+    options: AuthRequesterOptions<any>,
+  ): AuthRequest | undefined {
+    const { scopes } = request;
+    if (!scopes) {
+      return undefined;
+    }
+
+    return {
+      info: options.info,
+      triggerAuth: async () => {
+        const result = await options.authHandler(scopes);
+        request.resolve(result);
+      },
+      reject: () => {
+        const error = new Error('Login failed, rejected by user');
+        error.name = 'RejectedError';
+        request.reject(error);
+      },
+    };
+  }
+
+  handleAuthRequests(): Observable<AuthRequest[]> {
+    return this.requests$;
+  }
+
+  async showLoginPopup(options: LoginPopupOptions): Promise<any> {
+    return new Promise((resolve, reject) => {
+      const width = options.width || 500;
+      const height = options.height || 700;
+      const left = window.screen.width / 2 - width / 2;
+      const top = window.screen.height / 2 - height / 2;
+
+      const popup = window.open(
+        options.url,
+        options.name,
+        `menubar=no,location=no,resizable=no,scrollbars=no,status=no,width=${width},height=${height},top=${top},left=${left}`,
+      );
+
+      if (!popup || typeof popup.closed === 'undefined' || popup.closed) {
+        reject(new Error('Failed to open auth popup.'));
+        return;
+      }
+
+      const messageListener = (event: MessageEvent) => {
+        if (event.source !== popup) {
+          return;
+        }
+        if (event.origin !== options.origin) {
+          return;
+        }
+        const { data } = event;
+        if (data.type !== 'oauth-result') {
+          return;
+        }
+
+        if (data.payload.error) {
+          const error = new Error(data.payload.error.message);
+          error.name = data.payload.error.name;
+          // TODO: proper error type
+          // error.extra = data.payload.error.extra;
+          reject(error);
+        } else {
+          resolve(data.payload);
+        }
```

Comment:
the message contract isn't documented yet right?

Reply:
Nop